### DISCUSSION TITLE
Telescope ignores .env* files

### DISCUSSION THREAD

**Discussion by olrtg** [LunarVim/LunarVim#3770] [Q&A] [answered]
Hey guys! Don't know if this is a bug or not. I've been using LunarVim for like a year right now and I think this is a problem that I had since day one using telescope.

For some reason `.env` files are ignored (some screenshots for reference):

**`.env` or `.env.local` file is in the file tree:**
<img width="1552" alt="Screen Shot 2023-01-23 at 08 39 17" src="https://user-images.githubusercontent.com/20072509/214031254-dfd8797e-92ea-482e-b5b3-33fe90631936.png">

**Other "hidden" files (or dotfiles) are shown but no `.env` files:**
<img width="1552" alt="Screen Shot 2023-01-23 at 08 39 34" src="https://user-images.githubusercontent.com/20072509/214031286-6c11aaa6-c171-4cd8-842a-1da3b32272b8.png">

NOTE: I've tried using the `find_files` picker because I thought it was because the file is ignored by git but it's the same.

**Comment by olrtg** [comment]
UPDATE: after some quite a lot of investigation I discovered that telescope hides ignored files by default. Now I'll try to index ignored files but I need to create a list of ignored patterns (will check vscode for this).

Still, if you already figured this out please let me know.

**Comment by juicy-g** [comment]
These three commands work for me to find `.env.*` files:

```lua
:lua require('telescope.builtin').find_files({hidden=true})
```
```
Telescope find_files find_command=rg,--hidden,--files
```
```
Telescope find_files hidden=true
```

**Reply by olrtg** [reply]
Well, sadly this doesn't work for me still 😿. LunarVim has the hidden flag in `true`.

**Reply by martin-braun** [reply]
@juicy-g I have the same problem doing `:Telescope find_files hidden=true`. I can find `.env.example`, but I can't find `.env`. I also tried `find_command=rg` without luck. What's going on?

**Reply by juicy-g** [reply]
@martin-braun Weirdly enough now I can reproduce the problem which wasn't the case before. I am able to find hidden and git ignored files using:
```
Telescope find_files find_command=rg,--ignore,--hidden,--files,-u
```
Does it work for you?

`-u` is short for `--unrestricted` which means disable `.gitignore` handling.

**Reply by rendivs925** [reply]
it works thanks

**Reply by willisplummer** [reply]
Is there a way to unignore .env but leave everything else excluded?

**Reply by juicy-g** [reply]
I would try: `Telescope find_files find_command=rg,--ignore,--hidden,--files,--glob=.env`.

**Reply by willisplummer** [reply]
@juicy-g this will only include the .env file I think. I've been trying to find a way to include everything in cwd, hidden files, and .env but not the rest of my .gitignore... haven't really found a solution though - might not be doable

**Reply by juicy-g** [reply]
@willisplummer Ah ok. You're right. I didn't understand what you were trying to do. I found this discussion in the ripgrep Github: [https://github.com/BurntSushi/ripgrep/discussions/2428](https://github.com/BurntSushi/ripgrep/discussions/2428). 

Creating a `.rgignore` file in a project root directory with `!/.env` and then using `Telescope find_files find_command=rg,--ignore,--hidden,--files` seems to work for me.

**Reply by willisplummer** [reply]
Thanks yeah this is the best I've found. I was hoping that there was a way to configure this at the level of my nvim config so it applied to all projects but everything I've seen suggests that due to the order of priority, an exclude in .env will always take precedence unless you specify it in the folder's .rgignore

**Reply by raRaRa** [reply]
> Telescope find_files find_command=rg,--ignore,--hidden,--files,--glob=.env

Nice! Now I want to create a keymap for this. leader-s-e.

**Reply by juicy-g** [reply]
> > Telescope find_files find_command=rg,--ignore,--hidden,--files,--glob=.env
> 
> Nice! Now I want to create a keymap for this. leader-s-e.

@raRaRa That line doesn't really work. It will only show `.env` and nothing else. I left the comment here in the discussion for completeness.

**Reply by willisplummer** [reply, marked as answer]
found a solution to this that works for me -- I added a `.ignore` file in my home directory with the following:

```
!.env
!*/.env
```

the .ignore takes precedence over .gitignore in all child directories and now telescope includes .env in results.

I wanted to find a way to keep the .ignore file with my telescope config in ~/nvim/ and then pass it to telescope as an argument, but I couldn't get that to work. (e.g. `table.insert(vimgrep_arguments, "--ignore-file ~/nvim/.ignore")` -- see [this ripgrep comment](https://github.com/BurntSushi/ripgrep/discussions/2011#discussioncomment-1414366) that suggests you can set a path to an ignore

Hope this is helpful for others!

**Reply by olrtg** [reply]
@willisplummer this does what I wanted originally. Thank you so much!

**Reply by Barmanji** [reply]
Boi oh Boi, no matter how much chat gpt has improved, just a little help with real human makes it so easy. 
i wanted this for so long and now I got it! THANK YOU VERY MUCH! @willisplummer sir.

**Reply by tomaszferens** [reply]
@willisplummer Wow. The `--ignore-file` didn't work for me too! I would have never expected that the file has to be in home directory.

Thank you so much for this!

I added this script in my nvim `init.lua`:

```lua
-- Check and create .rgignore if it doesn't exist
local home = os.getenv "HOME"
local rgignore_path = home .. "/.rgignore"
local rgignore_file = io.open(rgignore_path, "r")

if not rgignore_file then
  rgignore_file = io.open(rgignore_path, "w")
  if rgignore_file then
    rgignore_file:write "!.env*\n"
    rgignore_file:close()
  end
else
  rgignore_file:close()
end
```
